feat: add multi file capabillity

moT01 opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

moT01 committed:
feat: add multi file capabillity

ojeytonwilliams committed:
chore: update react-monaco-editor

ojeytonwilliams committed:
feat: simple prototype for dynamic tabs

While this works for html challenges, it should dynamically generate the
tabs in a predetermined order from the available files.

ojeytonwilliams committed:
fix: prevent unnecessary re-renders

The editor should handle its own rendering.  By including a key that
changes as the users changes tabs forces the editor to re-render.  This
gets triggered as the user types, resetting the cursor to the start of
the editor.

Now the key only changes if the theme changes.

ojeytonwilliams committed:
feat: support multiple editor tabs

Generates models to handle multiple groups of code (html, js or css) and
tracks the editor state as user moves between tabs. The current model
is connected to the redux store via a key associated with the file in
the editor.

ojeytonwilliams committed:
fix: concat sources so that 'code' has them all

ojeytonwilliams commented:
I've been concentrating on the early HTML challenges, which it works nicely for.  It also handles JavaScript challenges partially.  If you click on the tab it shows the editor and everything works as expected, but it doesn't show by default (yet).

React challenges are not handled at all, though it should be relatively easy to add support for them.

ojeytonwilliams committed:
fix: relax challenge validation for multiple seeds

ojeytonwilliams committed:
feat: add simple challengeFile sorter

ojeytonwilliams committed:
feat: initialise editor with 'first' file

Since sort-files exists, the challengeFiles have an order. The editor
now displays the first file on mount.

ojeytonwilliams committed:
feat(client): handle jsx in multi-file editor

ojeytonwilliams committed:
fix(client): update editor when code is reset

ojeytonwilliams committed:
feat(parser): handle multi-file solutions

ojeytonwilliams committed:
feat(curriculum): validate multi-file solutions

ojeytonwilliams committed:
refactor: sort-files -> common utils

ojeytonwilliams committed:
refactor(multi) move preparation into curriculum

ojeytonwilliams commented:
@raisedadead  Ideally I think /client should depend on /utils and /curriculum, but /curriculum should only depend on /utils.  That way it should be easier to reason about and easier to drop in a new /client if and when that's desirable. 

So, as I've been going through this, I've been trying to disentangle /curriculum and /client a bit, putting various things in /utils. It's still a WIP, but my tentative plan is to cherry pick out those refactors into a separate PR, for ease of reviewing.  Does that sound reasonable?

ojeytonwilliams committed:
fix(multi): validate challenges correctly

ojeytonwilliams committed:
fix: remove redundant PropType

ojeytonwilliams committed:
fix: update editor on code reset

ojeytonwilliams committed:
feat: handle multi-file solutions

ojeytonwilliams committed:
WIP: temporary snapshot update

ojeytonwilliams committed:
WIP: modifies challenge to need multi-file soln

ojeytonwilliams committed:
fix(a11y): improve tab navigation for Editor

ojeytonwilliams committed:
feat: highlight currently selected editor tab

ojeytonwilliams committed:
WIP: use relevant test text

ojeytonwilliams commented:
Hey @moT01 to make things a little easier to manage, I've opened my own PR with these and a few more changes: https://github.com/freeCodeCamp/freeCodeCamp/pull/39223
